src/clp_ffi_js/ir/StructuredIrStreamReader.cpp

```diff
@@ -119,17 +125,23 @@ auto StructuredIrStreamReader::deserialize_stream() -> size_t {
                 )
         };
     }
-    m_timestamp_node_id = m_stream_reader_data_context->get_deserializer()
-                                  .get_ir_unit_handler()
-                                  .get_timestamp_node_id();
     m_stream_reader_data_context.reset(nullptr);
     return m_deserialized_log_events->size();
 }
 
 auto StructuredIrStreamReader::decode_range(size_t begin_idx, size_t end_idx, bool use_filter) const
         -> DecodedResultsTsType {
+    if (use_filter && false == m_filtered_log_event_map.has_value()) {
```

Comment:
Some of the logic is duplicated from the UnstructuredIrStreamReader. Do you think it would make sense to extract the common logic into the base class's `decode_range` implementation and let it call a private method of the subclasses? i.e.
1. Let's add a private API `StreamReader::decode_log_event` with a dummy implementation in `StreamReader` that throws "not implemented".
2. Let's override the method in both `StructuredStreamReader` and `UnstructuredStreamReader`.
3. Remove the overrides for `decode_range` in both `StructuredStreamReader` and `UnstructuredStreamReader`.
4. Implement `StreamReader::decode_range` which calls `this->decode_log_event()`.

@kirkrodrigues any objection for this Template Method pattern?

Reply:
I was thinking we could throw a Not Implemented error in the dummy implementation but forgot the cpp compiler should have caught such unimplemented abstract methods in subclasses at compile time. should make it abstract then

Reply:
> not straight forward to have a base StreamReader::decode_range() access the derived member variables

true. yup i think moving the the shared code to utils.hpp is a good idea. i imagine a templated helper function can be created if that's what you mean.